pkg/bpf/bpf_linux.go

```diff
@@ -42,6 +42,33 @@ func createMap(spec *ebpf.MapSpec, opts *ebpf.MapOptions) (*ebpf.Map, error) {
 		metrics.BPFSyscallDuration.WithLabelValues(metricOpCreate, metrics.Error2Outcome(err)).Observe(duration.End(err == nil).Total().Seconds())
 	}
 
+	// Found incompatible map. Open the map from a pin to add some more context to
+	// the error. Take care not to shadow err so it can be returned later.
+	if errors.Is(err, ebpf.ErrMapIncompatible) {
+		if opts == nil {
+			return nil, fmt.Errorf("empty MapOptions, can't open pin: %w", err)
+		}
+
+		path := path.Join(opts.PinPath, spec.Name)
+		m, err2 := ebpf.LoadPinnedMap(path, nil)
+		if err2 != nil {
+			return nil, fmt.Errorf("open pin of incompatible map: %w", err)
+		}
+		defer m.Close()
+
+		info, err2 := m.Info()
+		if err2 != nil {
+			return nil, fmt.Errorf("get info of incompatible map: %w", err)
+		}
+
+		return nil, fmt.Errorf(
```

Comment:
FWIW we could extend the library to return an more useful error. Or create a dedicated error type which contains the info needed.

Reply:
Yes, that might be better. Would like to merge this PR in the short term, though, so we can keep this here for now. Tracking this in https://github.com/cilium/ebpf/issues/1034.